### PR TITLE
reject zero length byte slices

### LINKED ISSUE #603: x/interchainstaking: Keeper.UnmarshalValidatorsRequest lacks check for unmarshalling zero length byte slices
## Summary of Bug
The referenced code lacks check for unmarshalling zero length byte slices https://github.com/quicksilver-zone/quicksilver/blob/bbf9ef90b82b0afde218e146d5c3c94da867c4dd/x/interchainstaking/keeper/keeper.go#L684-L691

Yet
The code right above for `(*Keeper).UnmarshalValidatorsResponse` has the validation per https://github.com/quicksilver-zone/quicksilver/blob/bbf9ef90b82b0afde218e146d5c3c94da867c4dd/x/interchainstaking/keeper/keeper.go#L670C78-L670C78
then down below too for `(*Keeper).UnmarshalValidator`

## Expected Behaviour
Some validation

## Version
13bf586baffd07ca4a2a7f093212455fafb5df27

____

/cc @elias-orijtech 

#### For Admin Use

- [x] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged/assigned

### PR DESCRIPTION
**Pull request by faddat**
## 1. Summary
Fixes #603 

## 2.Type of change

check to ensure that we don't unmarshal zero length byte slices

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## 3. Implementation details

Validate

### CODE REVIEW

File: x/interchainstaking/keeper/callbacks_test.go [hunk 1]
```diff
@@ -1142,7 +1143,13 @@ func TestValsetCallbackNilValidatorReqPagination(t *testing.T) {
...
 	data := []byte("\x12\"\n 00000000000000000000000000000000")
 	err := keeper.ValsetCallback(quicksilver.InterchainstakingKeeper, ctx, data, icqtypes.Query{ChainId: suite.chainB.ChainID})
-	suite.NoError(err)
+	if err != nil {
+		if err.Error() == "attempted to unmarshal zero length byte slice" {
+			fmt.Println("Successfully rejected zero lenfth byte slice")
+		} else {
+			t.FailNow()
+		}
+	}
```
Comment (ThanhNhann) [inline x/interchainstaking/keeper/callbacks_test.go:1152]: we should update these lines by `suite.Require().EqualError(err, "attempted to unmarshal zero length byte slice")`

File: x/interchainstaking/keeper/callbacks.go [hunk 2]
```diff
@@ -108,7 +108,7 @@ func RewardsCallback(k *Keeper, ctx sdk.Context, args []byte, query icqtypes.Que
 	// unmarshal request payload
 	rewardsQuery := distrtypes.QueryDelegationTotalRewardsRequest{}
 	if len(query.Request) == 0 {
-		return errors.New("attempted to unmarshal zero length byte slice (2)")
+		return errors.New("attempted to unmarshal zero length byte slice")
```
Comment (ThanhNhann) [inline x/interchainstaking/keeper/callbacks.go:111]: Plz keep these numbers for easy-to-distinguish errors from each other
And hey @joe-bowman, why don't we have this line `"attempted to unmarshal zero length byte slice (5)"`, I see we numbered from 1 to 9 but there is a leak of number 5

**Comment by joe-bowman** [comment]
I also agree with @ThanhNhann to keep the numbers. It makes debugging where this message comes from far easier!

**Comment by ThanhNhann** [comment]
hey @faddat, do you need help to complete this pr in case you are busy with another task?

File: x/interchainstaking/keeper/keeper.go [hunk 3]
```diff
@@ -670,37 +670,31 @@ func (k *Keeper) Rebalance(ctx sdk.Context, zone *types.Zone, epochNumber int64)
...
+	return validatorsRes, err
 }
 
 // UnmarshalValidatorsRequest attempts to umarshal  a byte slice into a QueryValidatorsRequest.
 func (k *Keeper) UnmarshalValidatorsRequest(data []byte) (stakingtypes.QueryValidatorsRequest, error) {
 	validatorsReq := stakingtypes.QueryValidatorsRequest{}
-	err := k.cdc.Unmarshal(data, &validatorsReq)
-	if err != nil {
-		return validatorsReq, err
+	if len(data) == 0 {
```
[the remark this answers is not in the document]
Comment (ThanhNhann) [reply x/interchainstaking/keeper/keeper.go:683]: It's like we don't need to add check the case `if len(data) == 0` here and can close #603 right @joe-bowman ?

[the remark this answers is not in the document]
Comment (ThanhNhann) [reply x/interchainstaking/keeper/keeper.go:683]: I think we should close this issue cuz it's not need to check the `len of data` here @joe-bowman

[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/keeper.go:683]: Agreed.

**Comment by joe-bowman** [comment]
Closing this, as the case being fixed is a genuine case where we should not reject zero length bytes.

**Pull request closed without merging**